src/engraving/dom/systemlock.cpp

```diff
@@ -0,0 +1,164 @@
+{
+    auto iter = m_systemLocks.find(mb);
+    return iter != m_systemLocks.end() ? iter->second : nullptr;
+}
+
+const SystemLock* SystemLocks::lockContaining(const MeasureBase* mb) const
+{
+    if (m_systemLocks.empty()) {
+        return nullptr;
+    }
+
+    auto iter = m_systemLocks.lower_bound(mb);
+    if (iter != m_systemLocks.begin()
+        && (iter == m_systemLocks.end() || mb->isBefore(iter->second->startMB()))) {
+        --iter;
+    }
+
+    const SystemLock* lock = iter->second;
+
+    return lock->contains(mb) ? lock : nullptr;
+}
+
+std::vector<const SystemLock*> SystemLocks::locksContainedInRange(const MeasureBase* start, const MeasureBase* end)
+{
+    std::vector<const SystemLock*> result;
+
+    for (auto& pair : m_systemLocks) {
+        const SystemLock* lock = pair.second;
+        if (start->isBeforeOrEqual(lock->startMB()) && lock->endMB()->isBeforeOrEqual(end)) {
+            result.push_back(lock);
+        }
+        if (lock->startMB()->isAfter(end)) {
+            break;
+        }
+    }
+
+    return result;
+}
+
+std::vector<const SystemLock*> SystemLocks::allLocks() const
+{
+    std::vector <const SystemLock* > locks;
+    locks.reserve(m_systemLocks.size());
+    for (auto& pair : m_systemLocks) {
+        locks.push_back(pair.second);
+    }
+    return locks;
+}
+
+#ifndef NDEBUG
+void SystemLocks::sanityCheck()
+{
+    for (auto iter = m_systemLocks.begin(); iter != m_systemLocks.end();) {
+        auto curIter = iter;
+        auto nextIter = ++iter;
+        if (nextIter == m_systemLocks.end()) {
+            break;
+        }
+
+        const MeasureBase* curMB = curIter->first;
+        const SystemLock* curSysLock = curIter->second;
+
+        DO_ASSERT(curSysLock->startMB() == curMB);
+
+        const MeasureBase* nextMB = nextIter->first;
+        const SystemLock* nextSysLock = nextIter->second;
+
+        DO_ASSERT(nextSysLock->startMB() == nextMB);
+        DO_ASSERT(curMB->isBefore(nextMB));
+        DO_ASSERT(curSysLock->endMB()->isBefore(nextSysLock->startMB()));
+    }
+}
+
+void SystemLocks::dump()
+{
+    for (auto& pair : m_systemLocks) {
+        const SystemLock* sl = pair.second;
+        LOGD() << "SystemLock --- Start measure: " << sl->startMB()->no() << ", End Measure: " << sl->endMB()->no();
+    }
+}
+
+#endif
+
+SystemLockIndicator::SystemLockIndicator(System* parent, const SystemLock* lock)
+    : m_systemLock(lock), EngravingItem(ElementType::SYSTEM_LOCK_INDICATOR, parent, ElementFlag::SYSTEM | ElementFlag::GENERATED) {}
+
+Font SystemLockIndicator::font() const
+{
+    Font font(configuration()->iconsFontFamily(), Font::Type::Icon);
+    static constexpr double STANDARD_POINT_SIZE = 12.0;
+    double scaling = spatium() / SPATIUM20;
+    font.setPointSizeF(STANDARD_POINT_SIZE * scaling);
+    return font;
+}
+
+void SystemLockIndicator::setSelected(bool v)
+{
+    EngravingItem::setSelected(v);
+    renderer()->layoutItem(this);
+    system()->page()->invalidateBspTree();

```

Comment:
This looks scary because we've never done it before... but on the other hand, I don't see what could go wrong. But when we start needing more hacks like this in the future, we should put some thought in whether we're on the right track with how we implement those elements that are half part of the notation, and half part of the UI.